[CORL-1365] Site Banning

nick-funk opened the pull request:
## What does this PR do?

Enable site specific banning for site moderators (optionally available to all other mods/users).

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds siteID's to the BanHistory, BanStatus, and BanUserInput.

## How do I test this PR?

- Enable SITE_MODERATOR feature flag
- Have multiple sites under org
- Ban people as various user roles (especially a site moderator).

Note: Site mods cannot un-ban a user

nick-funk committed:
Implement site specific banning

CORL-1365

wyattjoh commented on src/core/client/admin/components/UserStatus/UserStatusSitesListContainer.tsx:
```diff
@@ -0,0 +1,239 @@
+  const hasMore = !isRefetching && relay.hasMore();
+
+  const toggleShowSites = useCallback(() => {
+    setShowSites(!showSites);
```
```suggestion
    setShowSites((s) => !s);
```

wyattjoh commented on src/core/client/admin/components/UserStatus/UserStatusSitesListContainer.tsx:
```diff
@@ -0,0 +1,239 @@
+              {sites.map((site) => {
+                const enabled = siteIsVisible(site.id, viewerScopes.sites);
+                const isChecked = selectedIDsInput.value.includes(site.id);
+
```
Add a filter in a memo before the render to filter out sites that are not visible.

wyattjoh commented on src/core/client/admin/components/UserStatus/UserStatusSitesListQuery.tsx:
```diff
@@ -0,0 +1,52 @@
+            return <div>{error.message}</div>;
+          }
+
+          if (props) {
```
Suggest a `if (!props) { return null; }`

wyattjoh commented on src/core/server/graph/resolvers/UserStatus.ts:
```diff
@@ -53,10 +53,14 @@ export const UserStatus: Required<GQLUserStatusTypeResolver<
+      ...banStatus,
+      userID,
+    };
+  },
```
Match the format to `suspension`.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -2036,6 +2036,11 @@ type BanStatusHistory {
+  """
+  siteIDs are set when the ban was created to perform a site specific ban.
+  """
+  siteIDs: [String!]
```
```suggestion
  siteIDs: [ID!]
```

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -6945,6 +6955,11 @@ input BanUserInput {
+  """
+  site ID's to scope the ban to.
+  """
+  siteIDs: [String!]
```
```suggestion
  siteIDs: [ID!]
```

wyattjoh commented on src/core/server/models/user/user.ts:
```diff
@@ -1712,6 +1720,61 @@ export async function removeUserPremod(
+      tenantID,
+    },
+    {
+      $push: {
+        "status.ban.history": banHistory,
+      },
```
```suggestion
      $push: {
        "status.ban.history": banHistory,
      },
      $addToSet: {
        "status.ban.siteIDs": { $each: siteIDs }
      }
```

wyattjoh commented on src/core/server/models/user/user.ts:
```diff
@@ -1781,6 +1844,55 @@ export async function banUser(
+      tenantID,
+    },
+    {
+      $push: {
+        "status.ban.history": ban,
+      },
```
```suggestion
      $push: {
        "status.ban.history": ban,
      },
      $pull: {
        "status.ban.siteIDs": siteIDs,
      }
```

nick-funk replied:
In future, don't forget to include a `$in` or else it will try and delete the array of elements instead of each element and never pull anything from the array.

```
$pull: {
  "status.ban.siteIDs": { $in: siteIDs },
},
```

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -1057,35 +1059,65 @@ export async function ban(
-  const banStatus = consolidateUserBanStatus(targetUser.status.ban);
-  if (banStatus.active) {
-    throw new UserAlreadyBannedError();
+  let user: Readonly<User> | null = null;
```
```suggestion
  let user: Readonly<User>;
```

nick-funk committed:
Reduce memo usage where possible

CORL-1365

nick-funk committed:
Use safer toggle set lambda for setShowSites

CORL-1365

nick-funk committed:
Return null if props are unavailable in query

CORL-1365

nick-funk committed:
Simplify ban resolver

CORL-1365

nick-funk committed:
Store current banned site IDs on ban status

CORL-1365

nick-funk committed:
Make siteIDs of type ID on schema

CORL-1365

nick-funk committed:
Filter ban sites list based on scope visibility

CORL-1365

nick-funk committed:
Show in-stream site specific ban descriptions

CORL-1365

nick-funk committed:
Show number of banned sites in community tab

CORL-1471

nick-funk committed:
Allow filtering by site-banned in the community tab

CORL-1472

nick-funk committed:
Set sites in optimistic responses

CORL-1472

nick-funk committed:
Fix up optimistic responses and banned state calculation

Resolves the failing tests in previous commit.

CORL-1472

nick-funk committed:
Show "Suspend/Warn everywhere" when mod scopes enabled

CORL-1365

wyattjoh commented on src/core/server/app/middleware/tenant.ts:
```diff
@@ -1,16 +1,78 @@
+    storyID = "",
+    parentUrl = "",
+    siteID = "",
+  }: RequestQuery = req.query;
```
So there's currently a problem with how we're doing this. The requests where the site is important to the context is ones that request the `viewer`, but because these requests aren't the ones with the iframe parameters on, it means that these will not have the site attached to them, and can therefore never react to a site specific ban.

nick-funk committed:
Remove duplicate settings fragments

CORL-1365

nick-funk committed:
Float current ban status to top of account history

CORL-1474

nick-funk committed:
Show red colouring on floated account history bans

CORL-1474

nick-funk committed:
Handle admin, mod, site mod flows in ban modal

CORL-1475

nick-funk committed:
Check modScopes siteIDs property correctly

There was a case where the null coalescing operator
would cause the ban user flow in the back-end to
think an org mod was a site mod.

CORL-1365

wyattjoh committed:
fix: adapted to work without auth directive

wyattjoh committed:
fix: add support for parsing site details from referer header

wyattjoh committed:
feat: added helper to abstract check logic
